### DISCUSSION TITLE
Discounting future casflows with chainladder

### DISCUSSION THREAD

**Discussion by johalnes** [casact/chainladder-python#447] [General]
Hi,

Working on a demo to see if we can change from ResQ to this package together with notebooks or Streamlit. Mostly super simple to follow the documentation, and the package has developed nicely since last time I took a look some years ago! Really impressive 😄 

The one thing I somewhat struggle with is getting discount factors in an easy fashion. Using the package sample data
```python
sample = cl.load_sample('tail_sample')
sample['rbns'] = sample['incurred'] - sample['paid']
dev = cl.TailCurve(attachment_age=48,projection_period=0).fit_transform(sample)
```
Using Tailcurve for smoothing effect, and it's nice to get same dimensions as input triangle. 

My issue now is that I'll like to get the payment patterns from paid, and discount the RBNS with some interest curve, specified from EIOPA in my case. To do this I'll like to only look at remaining part of the payment pattern for each accident year. But using diff I get a dispatch errror from keepdims. 
<img width="548" alt="image" src="https://github.com/casact/chainladder-python/assets/8577145/e00f5821-42e4-4194-bf44-c045055273a7">

I also tried a bit with pct_change, but I think it's an error with [pct_chg](https://github.com/casact/chainladder-python/blob/1c0a3251526cf1dbc5a176794984eb7169575fce/chainladder/core/pandas.py#L481) should be [pct_change](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html?highlight=pct_change#pandas.DataFrame.pct_change).

I've added this to discussion instead of issue, beacuse I've got the impression that you have thought most thing over some times during this development and that maybe there is something clever that I may have missed? What would be your suggestion for discounting future cashflows? 

Please let me know if I should add diff and pct_change to issue, and I'll try to implement a fix 😄 

My solution is then to leave the safe space within chainladder and converts to pandas. 
```python
dev['payment_pattern']=(dev.ldf_['paid']/dev.cdf_['paid'])
payment_pattern=dev['payment_pattern'].to_frame().diff(axis=1).fillna(dev['payment_pattern'].to_frame())*np.flip(dev.nan_triangle)
payment_pattern.div(payment_pattern.sum(axis=1),axis=0)
```
|            |         12 |         24 |         36 |         48 |          60 |          72 |          84 |          96 |         108 |       120 |
|:-----------|-----------:|-----------:|-----------:|-----------:|------------:|------------:|------------:|------------:|------------:|----------:|
| 2000-01-01 | nan        | nan        | nan        | nan        | nan         | nan         | nan         | nan         | nan         | 1         |
| 2001-01-01 | nan        | nan        | nan        | nan        | nan         | nan         | nan         | nan         |   0.369578  | 0.630422  |
| 2002-01-01 | nan        | nan        | nan        | nan        | nan         | nan         | nan         |   0.365194  |   0.234611  | 0.400195  |
| 2003-01-01 | nan        | nan        | nan        | nan        | nan         | nan         |   0.358266  |   0.234357  |   0.150558  | 0.256819  |
| 2004-01-01 | nan        | nan        | nan        | nan        | nan         |   0.347421  |   0.233797  |   0.152937  |   0.0982507 | 0.167595  |
| 2005-01-01 | nan        | nan        | nan        | nan        |   0.330703  |   0.232528  |   0.15648   |   0.10236   |   0.0657589 | 0.112171  |
| 2006-01-01 | nan        | nan        | nan        |   0.305562 |   0.229653  |   0.161476  |   0.108665  |   0.0710827 |   0.0456655 | 0.0778955 |
| 2007-01-01 | nan        | nan        |   0.264291 |   0.224805 |   0.168958  |   0.118799  |   0.079946  |   0.0522962 |   0.0335965 | 0.0573084 |
| 2008-01-01 | nan        |   0.228429 |   0.203919 |   0.173453 |   0.130363  |   0.0916623 |   0.0616841 |   0.0403503 |   0.0259221 | 0.0442175 |
| 2009-01-01 |   0.290062 |   0.16217  |   0.14477  |   0.123141 |   0.0925495 |   0.0650745 |   0.0437918 |   0.0286462 |   0.0184031 | 0.0313917 |

And my next step would probably be flatting it into sql and join technical provisions by year and interest rate by development.

**Comment by jbogaardt** [comment]
Hi @johalnes . This sounds cool, Not having access to proprietary reserving software, its difficult for me to know the gaps that may exist in `chainladder`.  

I'm not familiar with European regs, so I could be off the mark here, but I typically use the `Chainladder.full_triangle_` to get patterns.

```python
import chainladder as cl

sample = cl.load_sample('tail_sample')
sample['rbns'] = sample['incurred'] - sample['paid']


pipe = cl.Pipeline([
    ('dev', cl.Development()),
    ('tail', cl.TailCurve(attachment_age=48, projection_period=0)),
    ('model', cl.Chainladder())])

pipe.fit(sample['paid'])

increm_future_paid = (pipe.named_steps.model.full_triangle_.cum_to_incr() - sample['paid'].cum_to_incr())
increm_future_paid_pct = increm_future_paid / increm_future_paid.sum('development')
```
As far as discounting goes, there is a [trend](https://chainladder-python.readthedocs.io/en/latest/user_guide/triangle.html#trend) method. It's not my favorite method because its a bit confusing to use. Still I don't think it will give you what you need because we don't have a good solution for the timing of cashflows in the tail.  `chainladder` assumes the cashflows at ultimate happen something like 150 years in the future which creates crazy results if you apply a discount factor. This is something we need to improve on.

For the sake of demonstration, (ignoring the cashflows at ultimate), the `trend` method could be used for discounting as follows: 
```python
discounted_increm_future_paid_pct  = increm_future_paid_pct.iloc[..., :-1].trend(
    0.05, axis='valuation', start=sample.valuation_date, end=sample.valuation.max())

discounted_increm_future_paid_pct
```

This may be way off base from your original question, if so, happy to take another stab at it with follow-up questions.
